docarray/array/array.py

```diff
@@ -13,34 +32,327 @@ class DocumentArray(
+        data back in each Document of the DocumentArray.
+
+        In unstacked mode DocumentArray behaves like a normal python list of Documents.
+        This means that you can call append, insert, index etc without problem.
+
+
+        You can still access DocumentArray fields in unstacked mode. This will return
+        a list of the fields of each Document in the array (or A DocumentArray if the
+        field contains Document and not type). This list/DocumentsArray will be
+        created on the fly and will not be stored in the DocumentArray. If you want to
+        do heavy operations on the fields of the DocumentArray you should use the stack
+        mode.
+
+        You can use setter as well to set the fields of the DocumentArray. This will
+        iterate over the list/DocumentsArray/Ndarray and set the field of each Document
+        accordingly.
+
+        EXAMPLE USAGE
+        .. code-block:: python
+            from docarray import Document, DocumentArray, Text
+            from docarray.typing import NdArray
+
+
+            class Image(Document):
+                tensor: NdArray[100]
+
+            batch = DocumentArray[Image]([Image(tensor=np.zeros((100))) \
+                for _ in range(10)])
+
+            batch.stack()
+            batch.unstack()
+            print(type(batch.tensor))
+            # >>> list
+            print(len(batch.tensor))
+            # >>> 10
+
+            batch.append(Image(tensor=np.zeros((100))))
+            # this work just fine
+
+            batch.tensor = np.ones((10, 100))
+            # this iterate over the ndarray and assign the row to each Document
+
+            print(batch[0].tensor[0])
+            # >>> 1
+
+        see {meth}`.stack` for more information on how to switch to stack mode
+        """
+        if self.is_stacked():
+
+            for field in list(self._columns.keys()):
+                # list needed here otherwise we are modifying the dict while iterating
+                del self._columns[field]
+
+            self._columns = None
+
+        return self
+
+    @contextmanager
+    def stacked_mode(self):
+        try:
+            yield self.stack()
+        finally:
+            self.unstack()
+
+    @contextmanager
+    def unstacked_mode(self):
+        try:
+            yield self.unstack()
+        finally:
+            self.stack()
+
+    def is_stacked(self) -> bool:
+        """
+        Return True if the document array is in stack mode
+        """
+        return self._columns is not None
+
+    def _column_fields(self) -> List[str]:
+        """
+        return the field store as columns
+        :return: the list of keys of the columns
+        """
+        if self.is_stacked() and self._columns is not None:
+            # need to repeat myself here bc of mypy
+            return list(self._columns.keys())
+        else:
+            return []
 
-        return _DocumenArrayTyped
+    append = _stacked_mode_blocker(list.append)
```

Comment:
Does this preserve docstrings?

Reply:
yes I believe I am using the context manager for decorator. I will check in the documentation. Actually I believe that we should add in the docstring at the end that it is not working is stacked mode